library-best-practices/glossary.md

```diff
@@ -0,0 +1,84 @@
+  dependency has changed in an incompatible way between the version supplied at
+  compile time and the version invoked at runtime. For example, a public method
+  may be removed from a class or a class may be made final. Linkage conflicts
+  detected at runtime manifest as ReflectiveOperationException,
+  `NoClassDefFoundError`, `NoSuchFieldException`, `MethodNotFoundException`,
+  `LinkageError`, or other related exceptions.
+  - Or, another perspective: In cases where binary compatibility and source
+    compatibility are the same, a linkage conflict is when compilation would
+    fail if the libraries in the classpath were all built together from their
+    originating source code, or when reflection would fail.
+  - Opposite: **Linkage-compatible**.
+  - Sub-type: **Static linkage conflict**: A linkage conflict caused by a direct
+    code reference (non-reflective).
+  - Sub-type: **Reflective linkage conflict**: A linkage conflict caused by a
+    reflective reference.
+- **Behavior conflict**: The class's implementation has changed in a way that
+  can break clients at runtime although all signatures remain compatible. For
+  example, if a method that has been returning mutable lists begins returning
+  immutable lists without updating its signature, dependents that mutate the
+  list will fail, possibly far away from the original call site. By contrast, a
+  change in return type from List to ImmutableList would be a linkage conflict.
+  - Opposite: **Behavior-compatible**.
+- **Version disagreement**: This is a class of failure reported by build systems
+  with strict version checking on. The main example is dependency convergence
+  failures when using Maven version enforcer, which checks that all branches of
+  a dependency tree agree on the version of every dependency. A failure of
+  dependency convergence doesn't guarantee a linkage or behavior conflict, and
+  the passing of a dependency convergence check doesn't guarantee there will be
+  no linkage or behavior conflicts. Consequently, version disagreement is a poor
+  proxy for real problems (linkage and behavior conflicts).
+  - Dependency convergence will fail even if the dependency tree only disagrees
+    at the level of a patch version (1.10.0 vs 1.10.1) and there is no linkage
+    or behavioral conflict.
+  - It is possible for dependency convergence to pass when a version of a
+    library is chosen at build time which is incompatible with another library
+    in the tree. This is possible by forcing Maven to pick the "wrong" version
+    using something like dependency management or dependency exclusion, which
+    silences the Maven enforcer.
+  - Opposite: **Version agreement**.
+- **Version alignment**: Said of the dependency tree of a Maven module. This
+  means that for any dependency of a module in that module's dependency tree,
+  all major Java build systems will select the same version of that dependency.
```

Comment:
Again, the definition should be a function of the Java classes, not the build tools.

Reply:
But the entire point of this concept relates to how build systems will form the classpath based on the dependency tree. If we take the build systems out of it, we have nothing.

Reply:
We will use it when I create a JLBP for BOMs. Should I wait to add it until then?

Reply:
We will use it when I create a JLBP for BOMs. Should I wait to add it until then?

Reply:
Probably We may end up defining it there instead. Hard to tell until we get there.

Reply:
Well the pertinent question is whether I should remove it from this PR in the meantime.

Reply:
I suggest removing it for now. We can always add it later when it's clearer.